expression/constant_propagation.go

```diff
@@ -138,10 +86,72 @@ func (s *propagateConstantSolver) propagateEQ() {
 	}
 }
 
-// validPropagateCond checks if the cond is an expression like [column op constant] and op is in the funNameMap.
-func (s *propagateConstantSolver) validPropagateCond(cond Expression, funNameMap map[string]bool) (*Column, *Constant) {
+// propagateColumnEQ propagates expressions like 'column A = column B' by adding extra filters
+// 'expression(..., column B, ...)' propagated from 'expression(..., column A, ...)' as long as:
+//
+//  1. The expression is deterministic
+//  2. The expression doesn't have any side effect
+//
+// e.g. For expression a = b and b = c and c = d and c < 1 , we can get extra a < 1 and b < 1 and d < 1.
+// However, for a = b and a < rand(), we cannot propagate a < rand() to b < rand() because rand() is non-deterministic
+//
+// This propagation may bring redundancies that we need to resolve later, for example:
```

Comment:
In the present planner, we can do this by adding a post-process after the logical optimization, before deriving the statistics and build index/table ranges.

In the new cascades planner, we can do this by adding a rule which only works on a `Filter` operator and remove the duplicated filters.

@bb7133 What's your opinion?

Reply:
Current ranger can spot these redundant filters and merge them?

Reply:
@eurekaka maybe we can do the enhancement in ranger?

Reply:
Seems like ranger would only simplify and merge these filters when they are considered to be index filters, if there is no index and no int primary key, these filters would be in PhysicalSelection as what they are originally; maybe we should extract the range simplification logic out of ranger, and this should be like the 'post-process' you mentioned above.